pkg/cli/image/manifest/source.go

```diff
@@ -0,0 +1,141 @@
+		// If ImageContentSourceFile is given, only add ImageContentSource from file, don't search cluster ICSP
+		if len(o.ImageContentSourcePolicyFile) != 0 {
+			return nil, nil
+		}
+		restConfig, err := f.ToRESTConfig()
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		icspClient, err := operatorv1alpha1client.NewForConfig(restConfig)
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		return icspClient.ImageContentSourcePolicies(), nil
+	}
+	return nil
+}
+
+// GetICSPs will lookup ICSPs from cluster. Since it's not a hard requirement to find ICSPs from cluster, GetICSPs logs errors rather than returning errors.
+func (o *ImageSourceOptions) GetICSPs(icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface) {
+	icsps, err := icspClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// may or may not have access to ICSPs in cluster
+		// don't error if can't access ICSPs
+		klog.V(4).Infof("did not access any ImageContentSourcePolicies in cluster: %v", err)
+	}
+	if len(icsps.Items) == 0 {
+		klog.V(4).Info("no ImageContentSourcePolicies found in cluster")
+	}
+	o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, icsps.Items...)
+}
+
+// GetImageSourcePrefixes returns list of 'registry/repository/name' options to try
+// 'registry/repository/name' determined from examining ImageContentSourcePolicy objects.
+// If a valid prefix is obtained from ICSP, return true
+func GetImageSourcePrefixes(icsps []operatorv1alpha1.ImageContentSourcePolicy, image string) (bool, []string, error) {
+	var imageSources []string
+	for _, icsp := range icsps {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		imageRef, err := reference.Parse(image)
+		if err != nil {
+			return false, imageSources, err
+		}
+		for _, rdm := range repoDigestMirrors {
+			rdmRef, err := reference.Parse(rdm.Source)
+			if err != nil {
+				return false, imageSources, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
+				klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+				sourceMatches = true
+			}
+			for _, m := range rdm.Mirrors {
+				if sourceMatches {
+					klog.V(2).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
+					imageSources = append(imageSources, m)
+				}
+				mRef, err := reference.Parse(m)
+				if err != nil {
+					return false, imageSources, err
+				}
+				if imageRef.AsRepository() == mRef.AsRepository() {
```

Comment:
Why do you need this?

Reply:
this is what allows
`oc adm release extract -a ~/pull-secret-local --icsp-file icsp.yaml --command openshift-install 4.6.0` to succeed, with this icsp.yaml, generated from `oc adm release mirror 4.6.0 --to localhost:5000/openshift/release`:
```
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: example
spec:
  repositoryDigestMirrors:
  - mirrors:
    - localhost:5000/openshift/release
    source: quay.io/openshift-release-dev/ocp-release
  - mirrors:
    - localhost:5000/openshift/release
    source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
```

Reply:
I've added comment explaining that GetImageSourcePrefixes function.

Reply:
With extract from release there is a call to this fn to find the correct release image, then to find imageSources for the image from which you'll extract (say, the `openshift-install` binary from the installer image) the first, for mirrored, you'd need to replace `quay.io/openshift-release-dev/ocp-release@sha256:something` with `localhost:5000/openshift/release@sha256:something`  _then_  to extract, have to match again for @sha256:component but this time you'll match the passed image with the mirrors instead of the source - since now the passed image will be the substituted release.

Reply:
I got rid of this, by adding the image-passed to the list of image sources